payone-adapter/src/main/java/com/commercetools/sunrise/payment/payone/PayonePaymentServiceProviderImpl.java

```diff
@@ -0,0 +1,51 @@
+package com.commercetools.sunrise.payment.payone;
+
+import com.commercetools.sunrise.payment.domain.CreatePaymentDataProvider;
+import com.commercetools.sunrise.payment.domain.PaymentServiceProvider;
+import com.commercetools.sunrise.payment.utils.PaymentPropertiesLoadingHelper;
+import io.sphere.sdk.payments.Payment;
+import io.sphere.sdk.payments.PaymentMethodInfo;
+import io.sphere.sdk.payments.PaymentStatus;
+
+import java.util.List;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Created by mgatz on 7/18/16.
+ */
+public class PayonePaymentServiceProviderImpl implements PaymentServiceProvider {
+
+    private PaymentPropertiesLoadingHelper propertiesLoadingHelper;
+
+    public PayonePaymentServiceProviderImpl() {
+        propertiesLoadingHelper = PaymentPropertiesLoadingHelper.createFromResource("methods/payone.properties");
+    }
+
+    @Override
+    public String getId() {
+        return propertiesLoadingHelper.getProperty("methods.interface");
+    }
+
+    @Override
+    public List<PaymentMethodInfo> getAvailablePaymentMethods() {
+        return propertiesLoadingHelper.getAvaiableMethodIds().stream().map(id -> propertiesLoadingHelper.getMethodInfo(id)).collect(Collectors.toList());
+    }
+
+    @Override
+    public Function<CreatePaymentDataProvider, Payment> provideCreatePaymentHandler(String methodId) {
```

Comment:
This is the PSP specific implementation which is called by the shop.
To return a Payment object (indeed this has been updated and returns a more data enriched wrapper object) directly would require blocking the thread until the call to CTP (and other possible stuff) is finished.
So the framework would decide to block the execution.

By returning the function it is the shop that is responsible for "applying" it. Either blocking or none blocking.

Btw: we only fetch the CompletionStage returned by the CTP SDK and apply more callbacks to it.